include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -5440,38 +5454,69 @@ namespace VectorTools
           .face_to_cell_index((line + 1) * (degree + 1) - 1, face);
 
       unsigned int associated_edge_dof_index = 0;
-      //       for (unsigned int face_idx = 0; face_idx < fe.dofs_per_face;
-      //       ++face_idx)
-      for (unsigned int line_idx = 0; line_idx < fe.dofs_per_line; ++line_idx)
+      for (unsigned int line_dof_idx = 0; line_dof_idx < fe.dofs_per_line;
+           ++line_dof_idx)
         {
-          // Assuming DoFs on a face are numbered in order by lines then faces.
+          // For each DoF associated with the (interior of) the line, we need
+          // to figure out which base element it belongs to and then if
+          // that's the correct base element. This is complicated by the
+          // fact that the FiniteElement class has functions that translate
+          // from face to cell, but not from edge to cell index systems. So
+          // we have to do that step by step.
+          //
+          // DoFs on a face in 3d are numbered in order by vertices then lines
+          // then faces.
           // i.e. line 0 has degree+1 dofs numbered 0,..,degree
           //      line 1 has degree+1 dofs numbered (degree+1),..,2*(degree+1)
           //      and so on.
-          const unsigned int face_idx = line * fe.dofs_per_line + line_idx;
+
+          const unsigned int face_dof_idx =
+            GeometryInfo<dim>::vertices_per_face * fe.dofs_per_vertex +
+            line * fe.dofs_per_line + line_dof_idx;
+
           // Note, assuming that the edge orientations are "standard"
           //       i.e. cell->line_orientation(line) = true.
-          const unsigned int cell_idx = fe.face_to_cell_index(face_idx, face);
-
-          // Check this cell_idx belongs to the correct base_element, component
-          // and line:
-          if (((dynamic_cast<const FESystem<dim> *>(&fe) != nullptr) &&
-               (fe.system_to_base_index(cell_idx).first == base_indices) &&
-               (lower_bound <= fe.system_to_base_index(cell_idx).second) &&
-               (fe.system_to_base_index(cell_idx).second <= upper_bound)) ||
-              (((dynamic_cast<const FE_Nedelec<dim> *>(&fe) != nullptr) ||
-                (dynamic_cast<const FE_NedelecSZ<dim> *>(&fe) != nullptr)) &&
-               (line * (degree + 1) <= face_idx) &&
-               (face_idx <= (line + 1) * (degree + 1) - 1)))
+          Assert(cell->line_orientation(line),
+                 ExcMessage("Edge orientation doesnot meet expectation."));
```

Comment:
Does it mean the method itself cannot work on non-standard meshes? I thought there are some tests, which prove the opposite (it was the purpose of creating FE_NedelecSZ...). However, unless you have non-homogeneous BCs, this may still work by coincidence...

Reply:
Good question. I don't think any of us really understands how this function works any more (maybe @winnifried is the exception). I'm thinking that this patch fixes a real and known bug and that it is a step forward. I would advocate for merging it.

@agrayver -- first, many thanks for actually looking at this patch! If you think that this breaks something, do you think you could create a test that you think worked before and now doesn't any more?

Reply:
@bangerth yes, this sounds good -- I will try to look through existing tests and see if this situation is not covered.

Reply:
@agrayver -- I ran all tests that had 'curl' or 'nedelec' in their name, and they seemed to work.